File: packages/evals/tasks/bench/observe/observe_file_uploads.ts
import { defineBenchTask } from "../../../framework/defineTask.js";
import { selectorsResolveToSameElement } from "../../../framework/observeSelectors.js";

export default defineBenchTask(
  { name: "observe_file_uploads" },
  async ({ debugUrl, sessionUrl, stagehand, page, logger }) => {
    try {
      await page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/file-uploads-3/");

      const { data: observations } = await stagehand.observe("find the file upload element");

      if (observations.length === 0) {
        return {
          _success: false,
          message: "observe returned no results",
          observations,
          debugUrl,
          sessionUrl,
          logs: logger.getLogs(),
        };
      }

      const expectedLocator = `xpath=/html/body/input`;

      const foundMatch = await selectorsResolveToSameElement(page, observations[0].selector, [
        expectedLocator,
      ]);

      return {
        _success: foundMatch,
        observations,
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    } catch (error) {
      return {
        _success: false,
        error: String(error),
        message: "returned selector does not resolve to same node as expected",
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    }
  },
);

File: packages/sdk-ruby/evals/tasks/observe/observe_file_uploads.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/observe/observe_file_uploads.ts

Evals.define_task("observe_file_uploads") do |t|
  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/file-uploads-3/")

  observations = t.stagehand.observe("find the file upload element").data

  if observations.empty?
    next {
      _success: false,
      message: "observe returned no results",
      observations: observations.map(&:to_wire),
    }
  end

  expected_locator = "xpath=/html/body/input"

  # v3 compares backendNodeIds; the v4 Locator exposes no node identity
  # so the same element-identity check is
  # re-expressed in-page: resolve the observed selector and the expected
  # selector and compare element references.
  matched = t.page.evaluate(<<~JS)
    (() => {
      const observedSelector = #{observations[0].selector.to_json};
      const candidateSelectors = #{[expected_locator].to_json};
      const resolve = (selector) => {
        const raw = selector.startsWith("xpath=") ? selector.slice("xpath=".length) : selector;
        if (raw.startsWith("/") || raw.startsWith("(")) {
          const result = document.evaluate(raw, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null);
          return result.singleNodeValue;
        }
        return document.querySelector(raw);
      };
      const observed = resolve(observedSelector);
      if (!observed) return null;
      for (const candidate of candidateSelectors) {
        if (resolve(candidate) === observed) return candidate;
      }
      return null;
    })()
  JS
  found_match = !matched.nil?

  { _success: found_match, observations: observations.map(&:to_wire) }
end
